comm.cpp

```diff
@@ -215,6 +252,11 @@ int Endpoint::read_msg(struct buffer *pbuf)
     pbuf->len = expected_size;
 
     return 1;
+
+invalid_msg:
+    rx_buf.len -= expected_size;
+    memmove(rx_buf.data, rx_buf.data + expected_size, rx_buf.len);
```

Comment:
humn...  I don't think this is the right approach. You are discarding the entire buffer when the start byte should be in the middle, which means we will lose a bit more frames until we recover.

You should start the "Find magic byte as the start byte" above again.

This function is becoming to be unmanageable and should be split in smaller parts. "check_crc()" is an obviously independent part.